Get open zwave ValueID working properly

dhylands opened the pull request:
Changes the internal representation from a pointer to being a home_id and id.

dhylands commented:
r? @julienw

julienw commented on openzwave-sys/openzwave-c/manager.cc:
```diff
@@ -85,108 +85,133 @@ void manager_set_poll_interval(Manager * manager, int32 interval, bool between_p
 
 GET_VALUE_FUNC(as_bool, bool* value) {
-  return manager->GetValueAsBool(*id, value);
+  ValueID vid(cvid->home_id, cvid->id);
```
nit: Maybe you could have done a GET_VALUE_FUNC_IMPL macro to help removing some repetition here. But do as you wish.

julienw commented on openzwave-sys/openzwave-c/notification.cc:
```diff
@@ -15,8 +15,8 @@ extern "C" {
   }
 
-  ValueID const * notification_get_value_id(const Notification * notification) {
-    return &notification->GetValueID();
```
when I look at this now I feel so newbie ;)

julienw commented on openzwave-sys/openzwave-c/value_classes/value_id.cc:
```diff
@@ -4,60 +4,69 @@
-uint8 value_id_get_node_id(ValueID * value) {
-  return value->GetNodeId();
+uint8 value_id_get_node_id(cValueID *cvid) {
+  ValueID vid(cvid->home_id, cvid->id);
```
that's where I'm not so sure: we need to instanciate a new object at each call... but I don't have a good and simple idea right now so let's do this for now.

dhylands replied:
I think that we can do a cleanup and get rid of cValueID and just use ValueID here.

What's important is that the rust side

```
#[repr(C)]
pub struct ValueID {
  pub home_id: uint32_t,
  pub id: uint64_t
}
```

match the C++ struct/class. So we could change the above (from openzwave-sys/src/value_classes/value_id.rs) to be:

```
#[repr(C)]
pub struct ValueID {
  pub id: uint32_t,
  pub id1: uint32_t,
  pub home_id: uint32_t
}
```

and then that ValueID would match the C++ ValueID and we could eliminate cValueID (since the C++ ValudID class has no base classes and no virtual members, it's essentially the same thing as a plain old struct).

But I think I'd leave that for a separate PR since what we have now works, even if it isn't quite as efficient as it could be.

julienw replied:
and what we have now is more future-proof as it does not depend on anything private.

julienw commented on openzwave-sys/src/value_classes/value_id.rs:
```diff
@@ -24,17 +26,21 @@ pub enum ValueType {
+#[repr(C)]
+pub struct ValueID {
+  pub home_id: uint32_t,
+  pub id: uint64_t
```
nit: usually I use directly rust's u32 and u64 as these libc types do not bring much value.

julienw replied:
and actually we use them in the definitions below :)

dhylands replied:
Old habits die hard...

julienw commented on src/manager.rs:
```diff
@@ -228,37 +229,43 @@ impl Manager {
+            extern_manager::manager_get_poll_intensity(self.ptr, &cvid)
         }
     }
 }
```
just wondering why you use `value_id_from_packed_id` here whereas for other APIs you created ValueID from the cpp wrapper ? The latter seems less verbose and more readable.

dhylands replied:
Going in a different direction. On one case we're going from Rust to C++ and in the other case we're going in to the oppostie direction.

julienw commented on src/node.rs:
```diff
@@ -1,9 +1,12 @@
-use std::{ fmt, ptr };
+use std::{ cmp, fmt, ptr };
+
+//use std::collections::HashSet;
```
nit: do we need this comment ?

dhylands replied:
Nope - just a carry over (i.e. missed cleanup)

julienw commented on src/node.rs:
```diff
@@ -112,10 +119,27 @@ impl Node {
         Some((class_name, class_version))
     }
+
+    pub fn cmp_by_id(&self, other: &Node) -> cmp::Ordering {
```
why do we need this ?

dhylands replied:
We probably don't any more. When the nodes were in a Vec I used it as an ordering function passed to sort. Now that we're using BTreeSet we don't need it any more.

julienw commented on src/node.rs:
```diff
@@ -112,10 +119,27 @@ impl Node {
+            Some(NodeBasic::RoutingSlave) => "Routing Slave",
+            _ => "unknown"
+        }
+    }
```
nit: I don't know if this works, but worth trying implementing the Display trait for the NodeBasic enum ? (I don't know if it's possible at all)

julienw commented on src/node.rs:
```diff
@@ -112,10 +119,27 @@ impl Node {
+            Some(NodeBasic::StaticController) => "Static Controller",
+            Some(NodeBasic::Slave) => "Slave",
+            Some(NodeBasic::RoutingSlave) => "Routing Slave",
+            _ => "unknown"
```
nit: "Unknown Basic" ?

julienw commented on src/node.rs:
```diff
@@ -112,10 +119,27 @@ impl Node {
 impl fmt::Debug for Node {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        // The Z-Wave network is a mesh, so don't print neighbors. You'll quite
+        // possibly wind up with infinite output.
```
maybe it's possible to display the ids only ? would using `{}` just work or would we need to iter over the neighbors' structure ?

dhylands replied:
I think that we could print just the node ids of the neighbors as long as we don't use {:?} on the neigbors vec

julienw commented on src/value_classes/value_id.rs:
```diff
@@ -22,7 +22,8 @@ impl<'a> ValueList<'a> {
-            extern_manager::get_value_list_selection_as_string(manager_ptr, self.value_id.ptr, &mut raw_string, rust_string_creator)
+            let cvid = extern_value_id::value_id_from_packed_id(self.value_id.home_id, self.value_id.id);
+            extern_manager::get_value_list_selection_as_string(manager_ptr, &cvid, &mut raw_string, rust_string_creator)
         };
```
same remark here: this seems more readable to have this conversion done in the C wrapper ?

julienw commented on src/value_classes/value_id.rs:
```diff
@@ -109,41 +105,75 @@ impl ValueID {
+            ValueType::ValueType_String => "String",
+            ValueType::ValueType_Button => "Button",
+            ValueType::ValueType_Raw => "Raw",
+        }
```
same nit about implementing Display if it's possible

julienw commented on src/value_classes/value_id.rs:
```diff
@@ -397,7 +460,7 @@ use std::cmp::{self, Ordering};
     fn eq(&self, other: &ValueID) -> bool {
-        unsafe { extern_value_id::value_id_eq(self.ptr, other.ptr) }
+        self.home_id == other.home_id && self.id == other.id
     }
```
maybe we can just use derive for PartialEq and Eq ?

dhylands replied:
Possibly. I noticed that PartialOrd calls self.eq and I wasn't sure if that would be implemented if we dreived. I'll try it.

julienw replied:
heh that's the point of deriving :) automatically implementing it

julienw commented on src/value_classes/value_id.rs:
```diff
@@ -405,7 +468,11 @@ impl cmp::Eq for ValueID {}
+        let is_less_than = unsafe {
+            let cvid_self = extern_value_id::value_id_from_packed_id(self.home_id, self.id);
+            let cvid_other = extern_value_id::value_id_from_packed_id(other.home_id, other.id);
+            extern_value_id::value_id_less_than(&cvid_self, &cvid_other)
```
I don't understand here: you're creating a C ValueID here, and then once again in the Cpp wrapper.

dhylands replied:
This is creating a cValueID. The C++ wrapper is creating a ValueID (the C++ object)

julienw commented on openzwave-sys/openzwave-c/value_classes/value_id.h:
```diff
@@ -8,29 +8,38 @@ typedef OpenZWave::ValueID ValueID;
 typedef ValueID::ValueGenre ValueGenre;
 typedef ValueID::ValueType ValueType;
 
+typedef struct cValueID {
```
nit: shouldn't this be called rustValueID? Because this is really the type we use in Rust ?

dhylands replied:
Right now we have 3 separate structures:

1 - ValueID from src/value_classes/ValueID.h. This is pure rust and can have impl methods applied to it

```
#[derive(Clone, Hash)]
pub struct ValueID {
    pub home_id: u32,
    pub id: u64
}
```

2 - cValueID from openzwave-sys/openzwave-c/value_classes/value_id.h 

```
typedef struct cValueID {
    uint32  home_id;
    uint64  id;

    cValueID(uint32 home_id, uint64 id)
      : home_id(home_id), id(id) {}
} cValueID;
```

with the rust side equivalent

```
#[repr(C)]
pub struct ValueID {
  pub home_id: uint32_t,
  pub id: uint64_t
}
```

from openzwave-sys/srv/value_id_classes/valid_id.rs

3 - ValueID class from the OpenZWave C++ code.

cValueID is used by the C code to convert the rust API into the C++ API. I think we can eliminate it entirely.

julienw replied:
ok I think I understood; I'll look at it once again with a clearer mind after this night.

dhylands commented:
I added a new commit which addresses the review comments and also eliminates cValueID altogether. I think that this simplifies things quite a bit.

If we decide to keep this, I'll squash the commit together.

dhylands commented:
I also discovered that we can add Display trait to C structs (ValueID). The NodeBasic stuff was a bit messier because it returns Option<NodeBasic> and not just a plain NodeBasic.

julienw commented on openzwave-sys/src/value_classes/value_id.rs:
```diff
@@ -24,17 +24,44 @@ pub enum ValueType {
+  pub id: u32,
+  pub id1: u32,
+  pub home_id: u32,
+}
```
wondering if it would be easier to implement a ValueID::from_packed that would call `value_id_from_packed_id` to make things easier for the callers ?

julienw replied:
also maybe you can derive Clone and Copy right now :)

dhylands replied:
I don't think that we need to from the internal ValueID (the above ValueID is the rust representation of the C++ ValueID class).
We already derive Clone and Hash for the rust ValueID

julienw replied:
yeah it's not a big deal but the rust book recommends to generally derive Clone and Copy when it's cheap like here. Again, not a big deal so I wouldn't block on this anyway

julienw commented on openzwave-sys/openzwave-c/value_classes/value_id.cc:
```diff
@@ -4,60 +4,60 @@
+                            uint8 command_class_id,
+                            uint8 instance,
+                            uint8 value_index,
+                            ValueType type) {
```
nit: indentation

julienw commented on src/manager.rs:
```diff
@@ -228,37 +241,43 @@ impl Manager {
 
     pub fn enable_poll_with_intensity(&mut self, value: &ValueID, intensity: u8) -> bool {
```
not for this PR, but tell me what you think: I think we should rename the Rust's ValueID just to Value. This should also reduce the confusion between openzwave's ValueID and our ValueID...

dhylands replied:
Then what about the rest of them (i.e. Node, Controller, Notification?) I think that having the rust one named after the C++ ones makes perfect sense and its a direct parallel.

julienw commented on src/manager.rs:
```diff
@@ -228,37 +241,43 @@ impl Manager {
     pub fn enable_poll_with_intensity(&mut self, value: &ValueID, intensity: u8) -> bool {
         unsafe {
-            extern_manager::manager_enable_poll_with_intensity(self.ptr, value.ptr, intensity)
+            let cvid = extern_value_id::value_id_from_packed_id(value.home_id, value.id);
```
for this PR or for later, as you wish, but as I suggested elsewhere, it would be clearer to have a shorter ExternValueID constructor.

julienw replied:
eg `CValueID::from_packed_id` :)

julienw replied:
it could also encapsulate the unsafe block (as we're quite sure this is not actually unsafe).

dhylands replied:
So you're talking about the "value_id_from_packed_id" name?
I think it needs to be value_id_xxx since its the name of a C function. Did you just want to call it "value_id"?

julienw commented on src/node.rs:
```diff
@@ -48,7 +50,7 @@ impl Node {
         get_node_security as get_security -> u8,
         is_node_zwave_plus as is_zwave_plus -> bool,
-        get_node_basic as get_basic -> u8,
+        //get_node_basic as get_basic -> u8,
```
nit: just remove this line :)

julienw commented on src/node.rs:
```diff
@@ -112,10 +129,42 @@ impl Node {
+            _ => "unknown"
+        }
+    }
+    */
```
I guess you can remove this now :)

julienw commented on src/node.rs:
```diff
@@ -74,6 +76,21 @@ impl Node {
+        let manager_ptr = unsafe { extern_manager::get() };
+        unsafe {
+            NodeBasic::from_u8(extern_manager::get_node_basic(manager_ptr, self.home_id, self.node_id))
+        }
```
nit: put the `unsafe {}` block only around the call to `get_node_basic`

julienw commented on src/node.rs:
```diff
@@ -112,10 +129,42 @@ impl Node {
+    pub fn from_node(node: &Node) -> Neighbor {
+        Neighbor { home_id: node.get_home_id(), node_id: node.get_id() }
+    }
 }
```
not really convinced that we need a dedicated struct for this... this is just a Node and you created this struct only to be able to easily format a debug string. But I think you can just add a `simple_debug` method to `Node` that would output `format!("Node {{ home_id: {}, node_id {} }}", self.home_id, self.node_id)` and call it.

julienw commented:
Ok, I'm fine with this new approach. I left some last comments and I think we'll be good to go. About the shorter constructor, this can be for later or now as you wish. The rename of Rust's ValueID to Value can wait for a later patch.

dhylands commented:
I squashed all the commits together, and I think I addressed all of the review comments.

I created a new function ValueID::ozw_vid which constructs a C++ ValueID from the rust ValueID and changed all of the call sites to use that. So the only place that now calls value_id_from_packed_id is the ozw_vid function.

I found a nice way to do the Neighbor thing using `itertools::free::join`

julienw commented on openzwave-sys/openzwave-c/value_classes/value_id.cc:
```diff
@@ -4,60 +4,60 @@
 bool value_id_less_than(ValueID * self, ValueID * other) {
-  return self < other;
+  return *self < *other;
 }
```
ooops, thanks

julienw commented:
landed in https://github.com/fxbox/openzwave-rust/commit/55b4f056f7ce1f5ff6ee1a744284ccceb94fabf4
